### PR TITLE
Bug in loc not raising KeyError with MultiIndex containing no longer used levels

### LINKED ISSUE #40235: BUG: reindex on empty Series raises ValueError: zero-size array to reduction operation
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
index = pd.MultiIndex.from_tuples([], names=["first", "second"])
s = pd.Series([], index=index, dtype=object)
print(f"s...\n{s}\n")
print(f"reindexed...\n{s.reindex(['a'], level='first')}")
```

prints
```
s...
Series([], dtype: object)
```
and then raises.

#### Problem description

reindex raises `ValueError: zero-size array to reduction operation maximum which has no identity` from `numpy/core/_methods.py`. Traceback:

<details>
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-37-935bb97993f2> in <module>
      3 s = pd.Series(range(len(index)), index=index, dtype=object)
      4 print(f"s...\n{s}\n")
----> 5 print(f"reindexed...\n{s.reindex(['a'], level='first')}")

~/.pyenv/versions/3.7.9/envs/covid-data-model/lib/python3.7/site-packages/pandas/core/series.py in reindex(self, index, **kwargs)
   4343     )
   4344     def reindex(self, index=None, **kwargs):
-> 4345         return super().reindex(index=index, **kwargs)
   4346 
   4347     def drop(

~/.pyenv/versions/3.7.9/envs/covid-data-model/lib/python3.7/site-packages/pandas/core/generic.py in reindex(self, *args, **kwargs)
   4810         # perform the reindex on the axes
   4811         return self._reindex_axes(
-> 4812             axes, level, limit, tolerance, method, fill_value, copy
   4813         ).__finalize__(self, method="reindex")
   4814 

~/.pyenv/versions/3.7.9/envs/covid-data-model/lib/python3.7/site-packages/pandas/core/generic.py in _reindex_axes(self, axes, level, limit, tolerance, method, fill_value, copy)
   4826             ax = self._get_axis(a)
   4827             new_index, indexer = ax.reindex(
-> 4828                 labels, level=level, limit=limit, tolerance=tolerance, method=method
   4829             )
   4830 

~/.pyenv/versions/3.7.9/envs/covid-data-model/lib/python3.7/site-packages/pandas/core/indexes/multi.py in reindex(self, target, method, level, limit, tolerance)
   2469                 target = ensure_index(target)
   2470             target, indexer, _ = self._join_level(
-> 2471                 target, level, how="right", return_indexers=True, keep_order=False
   2472             )
   2473         else:

~/.pyenv/versions/3.7.9/envs/covid-data-model/lib/python3.7/site-packages/pandas/core/indexes/base.py in _join_level(self, other, level, how, return_indexers, keep_order)
   3922             else:  # tie out the order with other
   3923                 if level == 0:  # outer most level, take the fast route
-> 3924                     ngroups = 1 + new_lev_codes.max()
   3925                     left_indexer, counts = libalgos.groupsort_indexer(
   3926                         new_lev_codes, ngroups

~/.pyenv/versions/3.7.9/envs/covid-data-model/lib/python3.7/site-packages/numpy/core/_methods.py in _amax(a, axis, out, keepdims, initial, where)
     37 def _amax(a, axis=None, out=None, keepdims=False,
     38           initial=_NoValue, where=True):
---> 39     return umr_maximum(a, axis, None, out, keepdims, initial, where)
     40 
     41 def _amin(a, axis=None, out=None, keepdims=False,

ValueError: zero-size array to reduction operation maximum which has no identity
</details>

#### Expected Output

I think it makes sense for `s.reindex(['a'], level='first')` to return an empty Series, as it does when `s` is not empty:

```python
index = pd.MultiIndex.from_tuples([('bar', 'one'),('foo', 'two')], names=["first", "second"])
s = pd.Series(range(len(index)), index=index, dtype=object)
print(f"s...\n{s}\n")
print(f"reindexed...\n{s.reindex(['a'], level='first')}")
```

prints
```
s...
first  second
bar    one       0
foo    two       1
dtype: object

reindexed...
Series([], dtype: object)
```

In 1.0.3 I could use df.loc[['label']] to get a slice without an exception when no data was found but that started raising KeyError when I upgraded to 1.2.3. I'm using reindex per https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#reindexing

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f2c8480af2f25efdbd803218b9d87980f416563e
python           : 3.7.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.19.0-11-amd64
Version          : #1 SMP Debian 4.19.146-1 (2020-09-17)
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.3
numpy            : 1.20.1
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 47.1.0
Cython           : None
pytest           : 6.2.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : 0.5.0
gcsfs            : None
matplotlib       : 3.2.1
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.1
sqlalchemy       : 1.3.23
tables           : None
tabulate         : 0.8.7
xarray           : 0.16.2
xlrd             : 1.2.0
xlwt             : None
numba            : 0.52.0

</details>

### LINKED ISSUE #41170: BUG: df.loc[[x], :] fails if df has zero rows
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
df = pd.DataFrame({"A": [12,23,34,45]}, index = [list("aabb"), [0,1,2,3]])
print(df)
print("- - -")
print(df.loc[df.A < 30, :].loc[["b"], :])  # empty as expected
print("- - -")
print(df.loc[df.A < 10, :].loc[["b"], :])  # raises ValueError

```
#### Complete Output of Code Sample

<details>

```python-traceback
      A
a 0  12
  1  23
b 2  34
  3  45
- - -
Empty DataFrame
Columns: [A]
Index: []
- - -
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 889, in __getitem__
    return self._getitem_tuple(key)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 1060, in _getitem_tuple
    return self._getitem_lowerdim(tup)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 791, in _getitem_lowerdim
    return self._getitem_nested_tuple(tup)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 865, in _getitem_nested_tuple
    obj = getattr(obj, self.name)._getitem_axis(key, axis=axis)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 1113, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 1053, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexing.py", line 1254, in _get_listlike_indexer
    indexer, keyarr = ax._convert_listlike_indexer(key)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexes\multi.py", line 2559, in _convert_listlike_indexer
    _, indexer = self.reindex(keyarr, level=level)
  File "C:\test\venv124\lib\site-packages\pandas\core\indexes\multi.py", line 2470, in reindex
    target, indexer, _ = self._join_level(
  File "C:\test\venv124\lib\site-packages\pandas\core\indexes\base.py", line 3924, in _join_level
    ngroups = 1 + new_lev_codes.max()
  File "C:\test\venv124\lib\site-packages\numpy\core\_methods.py", line 39, in _amax
    return umr_maximum(a, axis, None, out, keepdims, initial, where)
ValueError: zero-size array to reduction operation maximum which has no identity

```

</details>

#### Problem description

In both cases the ``df.loc[df.A...]`` returns a dataframe that doesn't contain any rows with an index value of "b".
Accordingly in the first case the result of ``.loc[["b"], :]`` is an empty dataframe, but in the second case a ValueError is raised. The difference between the cases is that in the first case ``df.loc[df.A...]`` returns a dataframe with some rows (though none with index value "b"), while in the second case ``df.loc[df.A...]`` returns a dataframe with zero rows. 
I think that shouldn't make a difference.

In the original code ``.loc[df.A...]`` and ``.loc[["b"], :]`` are not directly combined in one expression, but the first one creates a selection of rows of the dataframe, this selection is processed further, and during this another expression uses the second ``.loc``.

The traceback looks very similar to the one in #40235. Maybe both bugs have a common root cause.

#### Expected Output

``df.loc[df.A < 10, :].loc[["b"], :]`` returns an empty dataframe like ``df.loc[df.A < 30, :].loc[["b"], :]`` does.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.9.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
...

pandas           : 1.2.4
numpy            : 1.20.2
pytz             : 2021.1
dateutil         : 2.8.1
...

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #41170
- [x] closes #40235
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

stumbled across a bug in MultiIndex.reindex, which caused the ValueError from the op

### CODE REVIEW

**Comment by attack68** [comment]
Not sure if this is relevant to this PR but wanted to raise to the point that `KeyErrors` with MultiIndexes I think are dangerous because a MultiIndex cannot be easily reindexed, to permit a valid key.

For example if you wanted to reindex a MultiIndex and insert a single key, `a`, into one level of a multilevel index then it is ambiguous how we might do that, and in the exhaustive case of allowing `a` to all possible combinations of values of keys within the other levels you will quickly exhaust the memory due to the number of combinations.

I previously raised an issue where I thought the opposite (basing my opinion solely on the case of consistency with single indexes) but revised later after I realised it could be unworkable.

**Comment by phofl** [comment]
Yes, I think I remember this discussion, but this does not really apply here, since "b" is not in the Index at all, the same as 

```
df = pd.DataFrame({"A": [12,23,34,45]}, index = [list("aabb"), [0,1,2,3]])
df.loc[["c"], :]
```

I think the discussion was about something like:
```
df.loc[("a", 3), :]
```
where both elements are in the index, but the combination of both is not?

File: pandas/tests/indexes/multi/test_reindex.py [hunk 1]
```diff
@@ -104,3 +104,14 @@ def test_reindex_non_unique():
     msg = "cannot handle a non-unique multi-index!"
     with pytest.raises(ValueError, match=msg):
         a.reindex(new_idx)
+
+
+@pytest.mark.parametrize("values", [[["a"], ["x"]], [[], []]])
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_reindex.py:109]: this was user facing in a Series.reindex right? can you add the user facing part of this test as well.

Comment (phofl) [reply pandas/tests/indexes/multi/test_reindex.py:109]: Done, but is essentially the same

**Comment by attack68** [comment]
> Yes, I think I remember this discussion, but this does not really apply here, since "b" is not in the Index at all, the same as
> 
> ```
> df = pd.DataFrame({"A": [12,23,34,45]}, index = [list("aabb"), [0,1,2,3]])
> df.loc[["c"], :]
> ```
> 
> I think the discussion was about something like:
> 
> ```
> df.loc[("a", 3), :]
> ```
> 
> where both elements are in the index, but the combination of both is not?

But my point in this context is that if a user has some value he want to slice over that is *not* in df.index for a single index he has two options:

 - the index can reindexed (to include "c") which is the suggested route given the introduction of a KeyError.
 - the indexer arg must be altered to explicitly exclude missing values from df.index (i.e. remove "c")

Since MultiIndexes cannot be reindexed without ambiguity or without necessarily expanding the number of level combinations to fill memory (in this case adding "c" would return (c,0) (c,1) (c,2) (c,3) which is exponentially bad) then the user is restricted, in the case of MultiIndexes to:

 - the slice must be altered.

Which means that for each level the user must compare their indexer level args and exclude those that are not present.

I think this level of restriction will be more problematic than simply not returning values that don't exist?

**Comment by phofl** [comment]
Ah ok, did not remember this. I think this is a discussion we should have, but the pr is as you mentioned above probably not the right place, since this makes behavior more consistent and does not really change anything. Maybe ping on the other issue were the initial discussion was?

**Comment by attack68** [comment]
ok take a look here: #41358

**Pull request merged**